### PR TITLE
feat(test): Add smart test detection (issue #258)

### LINKED ISSUE #258: feat(test): Add smart test detection to `core test`
## Summary
Make `core test` smart - detect changed files and run only relevant tests.

## Commands
```bash
core test                    # Run tests for changed files only
core test --all              # Run all tests
core test --filter UserTest  # Run specific test
core test --coverage         # With coverage report
```

## Smart Detection
1. Check git diff for changed files
2. Map files to test files:
   - `src/User.php` → `tests/UserTest.php`
   - `internal/user.go` → `internal/user_test.go`
3. Run only relevant tests

## Downstream
Closes host-uk/core-agent#59

### PR DESCRIPTION
**Pull request by Snider**
## Summary
Implements smart test detection for `core test` command.

## Task Context
See `.core/task/issue/258/` for full specification and plan:
- [spec.md](https://github.com/Snider/core-1/blob/dev/.core/task/issue/258/spec.md)
- [plan.md](https://github.com/Snider/core-1/blob/dev/.core/task/issue/258/plan.md)

## Related
Closes #258

---
🤖 This PR is set up for AI agent implementation via fork workflow.

## Summary by CodeRabbit

* **Documentation**
  * Added specification and implementation plan for Smart Test Detection: describes phased rollout, CLI flags for controlling detection and test selection (all, filter, coverage, base), change-detection and test-mapping behavior, selective test execution and coverage handling, rename/delete/new-file rules, edge-case handling (no changes/no tests/mixed), and testing/verification guidance.

### CODE REVIEW

File: .core/task/issue/258/plan.md [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+1. Create `cmd/core/cmd/test.go`
+2. Register `test` command with Cobra
+3. Add flags: `--all`, `--filter`, `--coverage`
+
+## Phase 2: Change Detection  
+1. Run `git diff --name-only HEAD` to get changed files
+2. Filter for `.go` files (exclude `_test.go`)
+3. Map each file to its test file:
+   - `internal/foo/bar.go` → `internal/foo/bar_test.go`
+   - Skip if test file does not exist
```
[the remark this answers is not in the document]
Comment (Snider) [reply .core/task/issue/258/plan.md]: Fixed. Plan now includes Phase 2 steps 3-4: `git diff --name-status` for rename/delete/add detection, N:M test file discovery across package directories, and warnings for changed files without tests.

File: .core/task/issue/258/plan.md [hunk 2]
```diff
@@ -0,0 +1,33 @@
+# Implementation Plan: Issue 258
+
+## Phase 1: Command Structure
+1. Create `cmd/core/cmd/test.go`
+2. Register `test` command with Cobra
+3. Add flags: `--all`, `--filter`, `--coverage`
+
+## Phase 2: Change Detection  
+1. Run `git diff --name-only HEAD` to get changed files
```
[the remark this answers is not in the document]
Comment (Snider) [reply .core/task/issue/258/plan.md]: Fixed. Plan Phase 2 step 1 now specifies dual strategies: `git diff --name-only HEAD` + `--cached` for local development, `git diff --name-only origin/dev...HEAD` for CI/PR context. Auto-detects CI via env vars, with `--base` flag override.

File: .core/task/issue/258/plan.md [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+2. Filter for `.go` files (exclude `_test.go`)
+3. Map each file to its test file:
+   - `internal/foo/bar.go` → `internal/foo/bar_test.go`
+   - Skip if test file does not exist
+
+## Phase 3: Test Execution
+1. Build `go test` command with detected test files
+2. Pass through `--coverage` flag as `-cover`
+3. Pass through `--filter` as `-run` pattern
+4. Stream output to terminal
```
[the remark this answers is not in the document]
Comment (Snider) [reply .core/task/issue/258/plan.md]: Fixed. Phase 3 now explicitly reuses `runTest()` from `internal/cmd/test/cmd_runner.go`. No direct `go test` invocation — all execution goes through the existing infrastructure preserving env setup, output filtering, coverage parsing, and styling.

File: .core/task/issue/258/plan.md [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+
+## Phase 4: Edge Cases
+- No changed files → inform user, optionally run all
+- No matching test files → inform user
+- `--all` flag → skip detection, run `go test ./...`
```
[the remark this answers is not in the document]
Comment (Snider) [reply .core/task/issue/258/plan.md]: Fixed. The `--all` fallback now calls `runTest()` with `pkg="./..."` instead of raw `go test ./...`. All paths go through the existing `core test` infrastructure.

File: .core/task/issue/258/spec.md [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+core test --coverage         # With coverage report
+```
+
+## Acceptance Criteria
+- [ ] Detect changed `.go` files via `git diff --name-only`
+- [ ] Map source files to test files (`foo.go` → `foo_test.go`)
+- [ ] Run only relevant tests via `go test`
+- [ ] Support `--all` flag to run all tests
+- [ ] Support `--filter` flag for pattern matching
+- [ ] Support `--coverage` flag for coverage reports
```
[the remark this answers is not in the document]
Comment (Snider) [reply .core/task/issue/258/spec.md:28]: Fixed. Spec now explicitly states Go-only scope: the existing `core test` command requires `go.mod` and targets Go projects. Added a note that future language support (PHP, etc.) would be separate detection strategies, but this issue covers Go only.
